documentation/TROUBLESHOOTING.md

```diff
@@ -0,0 +1,343 @@
+
+Some graph break reasons are insurmountable to TorchDynamo, and can't be easily fixed.
+- calling into a C extension other than torch is invisible to torchdynamo, and could do arbitrary things without TorchDynamo being able to introduce necessary guards to ensure that the compiled program would be safe to reuse
+
+### Identifying the cause of a graph break
+
+To identify all graph breaks in a program and the associated reasons for the breaks, `torchdynamo.explain` can be used. This tool runs TorchDynamo on the supplied function and aggregates the graph breaks that are encountered. Here is an example usage:
+
+```py
+from typing import List
+import torch
+import torchdynamo
+
+def toy_example(a, b):
+    x = a / (torch.abs(a) + 1)
+    print("woo")
+    if b.sum() < 0:
+        b = b * -1
+    return x * b
+
+explained = torchdynamo.explain(toy_example, torch.randn(10), torch.randn(10))
+print(explained[0])
+
+"""
+Dynamo produced 3 graphs, with 2 graph break and 6 ops. 
+ Break reasons: 
+
+1. call_function BuiltinVariable(print) [ConstantVariable(str)] {} 
+   File "t2.py", line 16, in toy_example
+    print("woo")
+ 
+2. generic_jump 
+   File "t2.py", line 17, in toy_example
+    if b.sum() < 0:
+ """
+```
+
+To throw an error on the first graph break encountered, `nopython` mode can be used. This disables TorchDynamo's python fallback, and only succeeds if the entire program is convertible to a single graph. Example usage:
+
+```py
+@torchdynamo.optimize(<compiler>, nopython=True)
+def toy_example(a, b):
+   ...
+```
+
+## Diagnosing Runtime Errors
+Below is the TorchDynamo compiler stack. 
+
+<img src="./images/td_stack.png" width=800>
+
+At a high level, the TorchDynamo stack consists of a graph capture from Python code (TorchDynamo) and a backend compiler. In this example the backend compiler consists of backward graph tracing (AOTAutograd) and graph lowering (TorchInductor)*. Errors can occur in any component of the stack and will provide full stack traces, but there are backends which enable users to run each component of the stack without the others enabled to narrow down the exact component which is causing the error.
+
+There are some backend options which can enable you to determine which component is causing the error if you're unable to understand the error message that is generated. These are the following:
+
+- `"eager"`: only runs torchdynamo forward graph capture and then runs the captured graph with PyTorch. This provides an indication as to whether TorchDynamo is raising the error.
+
+- `"aot_nop"`: runs torchdynamo to capture a forward graph, and then AOTAutograd to trace the backward graph without any additional backend compiler steps. PyTorch eager will then be used to run the forward and backward graphs. This is useful to narrow down the issue to AOTAutograd.
```

Comment:
maybe this should be renamed to just `aot` and not `aot_nop`?

Reply:
In the performance dashboard we call this `aot_eager`, we should pick one and stick with it.

Reply:
Here - https://github.com/pytorch/torchdynamo/pull/1247

Reply:
Yes I went with `aot_eager` as well